opm/grid/GraphOfGridWrappers.cpp

```diff
@@ -208,201 +208,159 @@ void extendGIDtoRank(const GraphOfGrid<Dune::CpGrid>& gog,
+    exportedCells.resize(cc.size());
+    int addedExportsSize = 0;
+    for (int i=0; i<cc.size(); ++i)
+    {
+        exportedCells[i].reserve(sizesOfExport[i]);
+        addedExportsSize += sizesOfExport[i];
+    }
+    exportList.reserve(exportList.size()+addedExportsSize);
+    for (const auto& pWell : wellsToExport)
+    {
+        const auto& [begin, end, rank] = pWell;
+        // remember to skip the well's first cell which already is in the importList
+        for (auto pgID = begin; ++pgID!=end; )
+        {
+            using AttributeSet = Dune::cpgrid::CpGridData::AttributeSet;
+            exportList.emplace_back(*pgID, rank, AttributeSet::owner);
 
-    return exportedWells;
+            exportedCells[rank].push_back(*pgID);
+        }
+    }
+    std::sort(exportList.begin(), exportList.end());
+    return exportedCells;
 }
 
-std::vector<std::vector<int>> communicateExportedWells(
-    const std::vector<std::vector<std::vector<int>>>& exportedWells,
+std::vector<int> nonblockingCommunicateExportedCells(
+    const std::vector<std::vector<int>>& exportedCells,
     const Dune::cpgrid::CpGridDataTraits::Communication& cc,
     int root)
 {
     // send data from root
-    std::vector<std::vector<int>> result;
+    std::vector<int> result;
     if (cc.rank()==root)
     {
-        for (int i=0; i<cc.size(); ++i)
+        assert((int)exportedCells.size()==cc.size());
+        assert((int)exportedCells[root].size()==0);
+        // check send success for each well
+        std::vector<MPI_Request> requestSize(2*(cc.size()-1));
+        std::vector<int> nrWells(cc.size()-1); // store well sizes (can not send temporary)
+
+        for (int i=0; i<cc.size()-1; ++i)
         {
-            if (i!=root)
-            {
-                int numWells = exportedWells[i].size();
-                int totsize = numWells+1;
-                for (const auto& well : exportedWells[i])
-                {
-                    totsize += well.size();
-                }
-                // data: {N, size0, data0, size1, data1,... size(N-1), data(N-1)},
-                std::vector<int> commData;
-                commData.reserve(totsize);
-                commData.push_back(numWells);
-                for (const auto& well : exportedWells[i])
-                {
-                    commData.push_back(well.size());
-                    for (const auto& gID : well)
-                    {
-                        commData.push_back(gID);
-                    }
-                }
-                assert(totsize==(int)commData.size());
-                int tag = 37; // a random number
-                MPI_Send(&totsize, 1, MPI_INT, i, tag++, cc);
-                MPI_Send(commData.data(), totsize, MPI_INT, i, tag, cc);
-            }
+            int ii=i+(int)(i>=root); // ii takes values {0,...,mpisize-1} but skips root
+            int tag = 37; // a random number
+            nrWells[i]=exportedCells[ii].size();
```

Comment:
Sorry, now I have to nitpick on spaces. (I don't like people doing that :see_no_evil:)

Two style changes concerning white space in three lines. It would be really cool if we would have some consistency when using spaces (at least in consecutive lines).
I think our general rule (that vanished from the Wiki) is to have space next to operators. Certainly next to assignment operators.